### PR TITLE
Add back and modify signing root

### PR DESCRIPTION
**Pull request by 0xKiwi**
This PR brings back SelfSignedRoot from `spec-v0.6` of `prysm` as it was left behind during migration and updates it to the latest SSZ spec.

Part of #3 

Also:
`truncateLast()` to remove the last property in a struct (usually signature)
Renames `SelfSignedRoot` to `SigningRoot`

### CODE REVIEW

File: hash.go [hunk 1]
```diff
@@ -38,6 +39,57 @@ func TreeHash(val interface{}) ([32]byte, error) {
...
 
+// SigningRoot returns the signed message for the value passed after truncating
+// the `signature` field.
+func SigningRoot(val interface{}) ([32]byte, error) {
+	valObj := reflect.ValueOf(val)
+	kind := valObj.Kind()
+
+	switch {
+	case kind == reflect.Struct:
+		truncated, err := truncateLast(valObj.Type())
```
Comment (nisdas) [inline hash.go]: paths for both types (struct and pointer) are close to identical, they should be abstracted away into a common function

File: hash_test.go [hunk 2]
```diff
@@ -209,12 +275,29 @@ func runMerkleHashTests(t *testing.T, merkleHash func([][]byte) ([]byte, error))
 	}
 }
 
+func runSigningRootTests(t *testing.T, signingRoot func(val interface{}) ([32]byte, error)) {
+	for i, test := range signatureRootTests {
+		output1, err := signingRoot(test.val1)
```
Comment (nisdas) [inline hash_test.go]: you ignore both the errors here

File: hash_test.go [hunk 3]
```diff
@@ -18,6 +18,23 @@ type merkleHashTest struct {
 	output, error string
 }
 
+type signatureRootTest struct {
```
Comment (nisdas) [inline hash_test.go]: change the naming, the name is outdated. maybe `selfSignedRootTest` is better

File: hash_test.go [hunk 4]
```diff
@@ -18,6 +18,23 @@ type merkleHashTest struct {
...
+type blockHeader struct {
+	Slot              uint64
+	PreviousBlockRoot []byte
+	Signature         []byte
+}
+
+type truncateTestStruct struct {
+	Slot              uint64
+	StateRoot         []byte
+	PreviousBlockRoot []byte
```
Comment (nisdas) [inline hash_test.go]: instead of `previousBlockRoot` use a better field name, ex: `dummyField`, `ignoredField`, etc .

File: hash.go [hunk 5]
```diff
@@ -38,6 +39,29 @@ func TreeHash(val interface{}) ([32]byte, error) {
 	return paddedOutput, nil
 }
 
+// SigningRoot returns the signed message for the value passed after truncating
```
Comment (prestonvanloon) [inline hash.go]: This comment is lacking. Why does it truncate? What length does it truncate to? 
What is the value that I'm supposed to use this with?

File: ssz_utils_cache.go [hunk 6]
```diff
@@ -95,6 +95,22 @@ type field struct {
 	sszUtils *sszUtils
 }
 
+func truncateLast(typ reflect.Type) (fields []field, err error) {
```
Comment (prestonvanloon) [inline ssz_utils_cache.go]: truncate last what? Please add a comment

File: hash.go [hunk 7]
```diff
@@ -38,6 +39,30 @@ func TreeHash(val interface{}) ([32]byte, error) {
 	return paddedOutput, nil
 }
 
+// SigningRoot truncates the last prop of the struct passed in
```
Comment (nisdas) [inline hash.go]:
```suggestion
// SigningRoot truncates the last property of the struct passed in
```

**Comment by prestonvanloon** [comment]
Please add PR description

File: hash_test.go [hunk 8]
```diff
@@ -209,12 +275,35 @@ func runMerkleHashTests(t *testing.T, merkleHash func([][]byte) ([]byte, error))
...
+		output1, err := signingRoot(test.val1)
+		if err != nil {
+			t.Fatalf("could not get the signing root of test %d, value 1 %v", i, err)
+		}
+		output2, err := signingRoot(test.val2)
+		if err != nil {
+			t.Fatalf("could not get the signing root of test %d, value 2 %v", i, err)
+		}
+		// Check values have same result hash
+		if err == nil && !bytes.Equal(output1[:], output2[:]) {
```
Comment (nisdas) [inline hash_test.go]: don't need `if err == nil` here, it is redundant

File: hash_test.go [hunk 9]
```diff
@@ -209,12 +275,35 @@ func runMerkleHashTests(t *testing.T, merkleHash func([][]byte) ([]byte, error))
...
+}
+
 func TestHash(t *testing.T) {
 	runHashTests(t, func(val interface{}) ([32]byte, error) {
 		return TreeHash(val)
 	})
 }
 
+func TestSigningRoot(t *testing.T) {
+	runSigningRootTests(t, func(val interface{}) ([32]byte, error) {
```
Comment (prestonvanloon) [inline hash_test.go]: Please don't use this format, it is a bad pattern that we need to remove.

Comment (0xKiwi) [reply hash_test.go]: Can we maybe make a tracking issue to fix the tests in this repo?
What would the preferred format be?

Comment (nisdas) [reply hash_test.go]: just run the tests in this function instead of calling another function to run them

**Pull request merged**